File: dotnet/core/workspace/csharp/adapters/allors.workspace.adapters.remote/database/security/ResponseContext.cs
// <copyright file="RemoteResponseContext.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Remote
{
    using System.Collections.Generic;
    using System.Linq;

    internal class ResponseContext
    {
        private readonly DatabaseConnection database;

        internal ResponseContext(DatabaseConnection database)
        {
            this.database = database;

            this.MissingAccessControlIds = new HashSet<long>();
            this.MissingPermissionIds = new HashSet<long>();
        }

        internal HashSet<long> MissingAccessControlIds { get; }

        internal HashSet<long> MissingPermissionIds { get; }

        internal long[] CheckForMissingAccessControls(long[] value)
        {
            if (value == null)
            {
                return null;
            }

            foreach (var accessControlId in value.Where(v => !this.database.AccessControlById.ContainsKey(v)))
            {
                this.MissingAccessControlIds.Add(accessControlId);
            }

            return value;
        }

        internal long[] CheckForMissingPermissions(long[] value)
        {
            if (value == null)
            {
                return null;
            }

            foreach (var permissionId in value.Where(v => !this.database.Permissions.Contains(v)))
            {
                this.MissingPermissionIds.Add(permissionId);
            }

            return value;
        }
    }
}

File: typescript/libs/workspace/adapters/json/system/src/lib/runtime/security/ResponseContext.ts
import { IRange, enumerate } from '@allors/workspace/adapters/system';
import { Database } from '../Database';

export class ResponseContext {
  constructor(private readonly database: Database) {
    this.missingAccessControlIds = new Set();
    this.missingPermissionIds = new Set();
  }

  missingAccessControlIds: Set<number>;

  missingPermissionIds: Set<number>;

  checkForMissingAccessControls(value: IRange): IRange {
    for (const id of enumerate(value)) {
      if (!this.database.accessControlById.has(id)) {
        this.missingAccessControlIds.add(id);
      }
    }

    return value;
  }

  checkForMissingPermissions(value: IRange): IRange {
    for (const id of enumerate(value)) {
      if (!this.database.permissions.has(id)) {
        this.missingPermissionIds.add(id);
      }
    }

    return value;
  }
}
